src/control/cmd/dmg/system_test.go

```diff
@@ -43,39 +43,65 @@ func TestDmg_SystemCommands(t *testing.T) {
 			"system query with no arguments",
 			"system query",
 			strings.Join([]string{
-				printRequest(t, &control.SystemQueryReq{
-					Ranks: []Rank{},
-				}),
+				printRequest(t, &control.SystemQueryReq{}),
 			}, " "),
 			nil,
 		},
 		{
 			"system query with single rank",
 			"system query --ranks 0",
 			strings.Join([]string{
-				printRequest(t, &control.SystemQueryReq{
-					Ranks: []Rank{0},
-				}),
+				`*control.SystemQueryReq-{"HostList":null,"Ranks":"0","Hosts":""}`,
 			}, " "),
 			nil,
 		},
 		{
 			"system query with multiple ranks",
-			"system query --ranks 0,1,4",
+			"system query --ranks 0,2,4-8",
 			strings.Join([]string{
-				printRequest(t, &control.SystemQueryReq{
-					Ranks: []Rank{0, 1, 4},
-				}),
+				`*control.SystemQueryReq-{"HostList":null,"Ranks":"0,2,4-8","Hosts":""}`,
```

Comment:
Yeah, I see that I wasn't clear in my feedback, sorry about that. What I was trying to indicate is that I think it would be less confusing and error-prone if we adopted the convention that we only use "*List" variable names for actual lists (slices or Host/RankList types). I have made the mistake of checking the length of a *List variable, forgetting that it's just a string, and the compiler is perfectly happy to let me make this mistake.

With regard to API inputs for host/rank lists, we have three options as I see it:
1. Accept string representations (in ranged or comma-separated format) and parse them in the API
2. Accept HostList or RankList formats
3. Accept slices

Of these three, I think 3) is the worst option. In particular for the ranks, we'd have to choose between slices of strings, slices of uint32, or slices of system.Rank. Too much ambiguity IMO.

The first option is appealing in that it centralizes responsibility for parsing the lists in the API, reducing the amount of parsing that has to be done out in the tool space. On the other hand, it opens the door for imprecision about what's being supplied to the API functions.

I think that leaves us with the middle option as being the best compromise between convenience and precision. We know that the hostlist is a valid hostlist by the time we're looking at it in the API, and the user interface code can deal with validating it and letting the user know how to fix things if validation fails.

So, all that having been stated, what about the protobuf messages? Right now we're using this convention to represent rank lists:
```
repeated uint32 ranks = 1;
```

This made sense before we had the `RankSet` type that Tom added. I wonder if we should consider switching these over to string fields to reduce the size of messages (imagine a RankSet that encompasses the entire A21 system, that's 2048 ranks and the message would need 64kib to represent the ranks vs. 8 bytes for "[0-2048]"). Premature optimization? Not sure, but worth thinking about.

Reply:
if we put *Set type in the request structures then we need to convert in dmg package then serialise back to string when issuing Protobuf request in the control module, it therefore makes sense to retain string representation of set in dmg and lib/control packages as much as possible and only unpack into *Set types when we need to.

Reply:
I think passing the string format in the proto message is a good idea, no issues with that.

> if we put *Set type in the request structures then we need to convert in dmg package then serialise back to string when issuing Protobuf request in the control module, it therefore makes sense to retain string representation of set in dmg and lib/control packages as much as possible and only unpack into *Set types when we need to.

@tanabarr but we're doing the same series of string->set->string conversions anyway to validate the input string, so I'm not sure it gains us anything to pass in the string format to the API. I think it adds some clarity to the API to pass in the *Set types.

Reply:
given we are now passing string set representation in the proto messages I don't really see the benefit of changing just the API to store *Set types, @mjmac could you please make an executive decision on this

Reply:
in the interest of progress I will make the requested change

Reply:
Thanks for making this change. I think it's the right move in the end. The details of what we do in the protobuf messages aren't relevant to the API surface we expose. The point of having separate types is to decouple those things.